### PR TITLE
chore: Add balance check before inserting MsgWrappedCreateValidator into the epoch message queue

### LINKED ISSUE #317: Lack of balance check before inserting MsgWrappedCreateValidator into the message queue
This issue is found in our Discord discussion. Due to the lack of balance check, a `MsgWrappedCreateValidator` might fail when execution at the end of an epoch. We should add a check before inserting it into the message queue.

### PR DESCRIPTION
**Pull request by gitferry**
This PR fixed #317 and added relevant fuzz tests. To support the check of delegator balance, this PR added the `BankKeeper` to the epoching module

### CODE REVIEW

File: x/epoching/keeper/staking_functions.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/babylonchain/babylon/x/epoching/types"
 )
 
 // CheckMsgCreateValidator performs stateless checks on a given `MsgCreateValidator` message
```
Comment (SebastianElvis) [inline x/epoching/keeper/staking_functions.go:12]:
```suggestion
// CheckMsgCreateValidator performs checks on a given `MsgCreateValidator` message
```

Now this function does not only perform stateless checks anymore

File: x/epoching/keeper/staking_functions.go [hunk 2]
```diff
@@ -89,9 +91,15 @@ func (k Keeper) CheckMsgCreateValidator(ctx sdk.Context, msg *stakingtypes.MsgCr
...
 	// sanity check on delegator address
-	if _, err := sdk.AccAddressFromBech32(msg.DelegatorAddress); err != nil {
+	delegatorAddr, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
 		return err
 	}
 
+	balance := k.bk.GetBalance(ctx, delegatorAddr, msg.Value.GetDenom())
+	if msg.Value.IsGTE(balance) {
+		return types.ErrInsufficientBalance
```
Comment (SebastianElvis) [inline x/epoching/keeper/staking_functions.go:101]: Is there any error that we can reuse from `x/bank` module?

File: app/app.go [hunk 3]
```diff
@@ -346,7 +350,7 @@ func NewBabylonApp(
 
 	// NOTE: the epoching module has to be set before the chekpointing module, as the checkpointing module will have access to the epoching module
 	epochingKeeper := epochingkeeper.NewKeeper(
-		appCodec, keys[epochingtypes.StoreKey], keys[epochingtypes.StoreKey], app.GetSubspace(epochingtypes.ModuleName), &app.StakingKeeper,
+		appCodec, keys[epochingtypes.StoreKey], keys[epochingtypes.StoreKey], app.GetSubspace(epochingtypes.ModuleName), app.BankKeeper, &app.StakingKeeper,
```
Comment (SebastianElvis) [inline app/app.go:353]: Should this be
```suggestion
		appCodec, keys[epochingtypes.StoreKey], keys[epochingtypes.StoreKey], app.GetSubspace(epochingtypes.ModuleName), &app.BankKeeper, &app.StakingKeeper,
```

?

File: x/checkpointing/keeper/msg_server_test.go [hunk 4]
```diff
@@ -97,6 +98,43 @@ func FuzzWrappedCreateValidator_InsufficientTokens(f *testing.F) {
 	})
 }
 
+// FuzzWrappedCreateValidator_InsufficientBalance tests adding a new validator
+// but the delegator has insufficient balance to perform delegating
+func FuzzWrappedCreateValidator_InsufficientBalance(f *testing.F) {
```
Comment (SebastianElvis) [inline x/checkpointing/keeper/msg_server_test.go:103]: Another thing we can do is to move tests on `MsgCreateValidator` from Cosmos SDK to here. Wdyt?

**Review by SebastianElvis — APPROVED** [review]
Lgtm! Just some minor comments

File: x/epoching/keeper/staking_functions.go (hunk 2 above)
Comment (gitferry) [reply x/epoching/keeper/staking_functions.go:101]: No. I suppose it's up to the user of the bank module to determine whether the balance is sufficient or not

File: app/app.go (hunk 3 above)
Comment (gitferry) [reply app/app.go:353]: I was wondering the same thing that why the staking keeper is passed as a pointer but not the bank keeper, but it turned out it should be as it is here...

File: x/checkpointing/keeper/msg_server_test.go (hunk 4 above)
Comment (gitferry) [reply x/checkpointing/keeper/msg_server_test.go:103]: I definitely agree but I couldn't find any tests in cosmos-sdk that do stateful tests like us here

**Review by vitsalis — APPROVED** [review]
Nice work!

**Pull request merged**